pyvista/utilities/reader.py

```diff
@@ -117,33 +120,56 @@ class BaseReader:
 
     _class_reader: Any = None
 
-    def __init__(self, filename):
-        """Initialize Reader by setting filename."""
+    def __init__(self, path):
```

Comment:
I like this better, but it will be a breaking change if anyone is using it with a keyword arg.

Reply:
> I like this better, but it will be a breaking change if anyone is using it with a keyword arg.

We could turn this into a `@classmethod`. That's the recommended approach for all Liskov violations.

Reply:
@adam-grant-hendry perhaps I'm missing your point, but how would that help? If I understood @MatthewFlamm's point correctly, the problem is that if someone did `Reader(filename='./...')` this will no longer work, because they would have to change that to `Reader(path='./...')`.

Actually, this is `BaseReader`, right? Presumably all concrete child classes override `__init__` so end users should probably not be using this param directly. Unless the parent class is concrete (I haven't checked). But this would only shift the blame from renaming the kwarg in concrete child classes.

Reply:
the `__init__` in `BaseReader` is not abstract and from skimming the file, most child classes do not override it. Most child classes don't do anything other than set their `_class_reader` attribute.

Reply:
Thanks @njneeteson. I did skim the diff in this PR and I thought I saw `__init__`s in some child, but then I must have misread. In that case I understand the name change in the base.

Reply:
> the `__init__` in `BaseReader` is not abstract and from skimming the file, most child classes do not override it. Most child classes don't do anything other than set their `_class_reader` attribute.

True, though I've made the class itself an abstract class just to let users know that you're not supposed to instantiate a `BaseReader` (tried that in testing and I think a non-developer would be confused by the error).

---

Back to the root issue:

> I like this better, but it will be a breaking change if anyone is using it with a keyword arg.

Correct me if I'm wrong, but renaming the `arg` shouldn't have any effect other than a doc change. There's no keyword arguments in this `__init__`.

Reply:
@akaszynski all arguments are also keyword arguments unless explicitly defined to be positional-only, see https://peps.python.org/pep-0570/#specification as of Python 3.8.

Reply:
> @akaszynski all arguments are also keyword arguments unless explicitly defined to be positional-only, see https://peps.python.org/pep-0570/#specification as of Python 3.8.

Learn something every day!

To do this right, we'd have to rename all subclassed `filename` to `path`, and pass `**kwargs` and then remove any **kwargs.

IMO, I think the change is small enough and early enough to not necessitate passing `**kwargs` as I doubt many will use `filename=`. I'm willing on implementing it though if you disagree.

Reply:
I'm okay with the hard breaking change it would be unusual in this case that it would be used as a keyword arg with only one param).

Reply:
I don't have a strong opinion either way. But if we break here, we should probably still rename all the subclasses as well (this is probably also what you meant, but it's better to be explicit).

Reply:
The aliasing kwarg thing seems kludgy. PyVista is at version 0.34? It's reasonable IMO for names of positional parameters to change.

If we're getting onto this subject, `filename` or `path` or whatever it is should probably be a `@property` and the contents of `_set_filename` (or `_set_filename_or_directory`) should be in the filename/path `.setter`, not a separate private method (or at least that private method should get called in the path `setter`). I'm not sure it makes sense for `filename` to be implied to be a public attribute but the method that actually propagates the filename or path to the reader object is implied to be private (leading underscore).

Reply:
> However, it would probably be better if filename and directory weren't separate properties and instead there was just a path property with a setter that calls something similar to `_set_filename_or_directory` - but just use path for the `SetFileName` or `SetDirectory` methods depending on what they are. IMO it's a bit confusing to have two separate properties that kind of do the same thing but only one is "active" at a time and it kind of precludes using a setter properly.

I agree with this. We should just call it `path` (unless someone can come up with something better).

Reply:
> However, it would probably be better if filename and directory weren't separate properties and instead there was just a path property with a setter that calls something similar to _set_filename_or_directory - but just use path for the SetFileName or SetDirectory methods depending on what they are.

We should align on what that looks like (maybe a simpler `set_path` method or `path` attribute with setter/getter?), but this PR has already overstepped its original intention. So let's get that in another PR.

Reply:
> @adam-grant-hendry perhaps I'm missing your point, but how would that help? If I understood @MatthewFlamm's point correctly, the problem is that if someone did `Reader(filename='./...')` this will no longer work, because they would have to change that to `Reader(path='./...')`.
> 
> Actually, this is `BaseReader`, right? Presumably all concrete child classes override `__init__` so end users should probably not be using this param directly. Unless the parent class is concrete (I haven't checked). But this would only shift the blame to renaming the kwarg in concrete child classes.

Classmethods are used to create alternate constructors in python. Take the python datetime module. It has many methods beginning with `from...` with a `cls` parameter. These are alternate constructors and are decorated with the `@classmethod` decorator. In addition, I would turn `filename` into an `@property`. This is exactly what properties are for in python. Python is a "consenting adults" language, meaning all variables are public and we don't require getters or setters ahead of time. Instead, we use them in these exact situations.

Here is a great talk from the core developer Raymond Hettinger on the subject (start at 23:42):

[Hettinger, Raymond. "Python's Class Development Toolkit". PyCon US 2013](https://www.youtube.com/watch?v=HTLu2DFOdTg)

What you can do is create a `from_folder` method and then call `__init__` from there. 

I would do either:

```python
class BaseReader:
    ...
    def __init__(self, filename):
        ...

    ...

    def from_folder(cls, path):
        # This is for readers that use directories...
```

and use `from_folder` in my readers that use directories, or turn `filename` into a property

```python
class BaseReader:
    ...
    def __init__(self, filename):
        ...

    ...

    @property
    def filename(self):
        return self.filename

    @filename.setter
    def filename(self, path):
        if os.path.isfile(path):
            self.filename = path
        elif os.path.isdir(path):
            self.dirname = path
            self.filename = None
        else:
            raise ScreamBloodyMurderError("What have you done?!!!")
```

Keyword arguments are also a reasonable solution, but I agree with @njneeteson it is also kludgy. The classmethod approach prevents an explosion of keyword arguments (particularly for alternate constructors that require many keyword arguments).

Reply:
I'm not a big fan of the class method solution but I like what you did in the second code block example. That's more or less what I was suggesting except rename `filename` to `path`. Just one path setter for all subclasses that detects whether the path is a filename or dirname and acts appropriately. 

If certain child classes need some other functionality when path is set then they can override the setter but it should be very rare that needs to be done.

Reply:
+1 for

```py
            raise UnholyHellrror("What have you done?!!!")
```

Reply:
I finally settled on `path` in 2c01008c8b034adc97d87c111fb7a39d4fdb71cd.